### DISCUSSION TITLE
Heroku "H18 - Server Request Interrupted" error (503) when a request exceeds the bodyLimit (413)

### DISCUSSION THREAD

**Discussion by josdejong** [fastify/fastify#4919] [Q&A] [answered]
I have a Fastify backend configured with the default `bodyLimit` of 1MB. When I run the backend locally and I send a PUT request with a large body exceeding 1MB, an error with status `413 Payload too large` is returned, which is as expected. However, when I deploy the backend on Heroku and send the same request, I get back a `503 Service Unavailable` error, and the Heroku logs report an [H18 - Server Request Interrupted](https://devcenter.heroku.com/articles/error-codes#h18-server-request-interrupted) error. This error is described as:

> An H18 signifies that the socket connected, and some data was sent; The error occurs in cases where the socket was destroyed before sending a complete response, or if the server responds with data before reading the entire body of the incoming request.

My assumption is that Fastify stops reading the request body as soon as the limit is reached, triggering this H18 error in Heroku, but I'm not sure at all.

Any idea on why this happens and how to solve this?

**Comment by Uzlopak** [comment, marked as answer]
I think you are correct with your assessment. But it seems to be more an issue with heroku than fastify.

**Reply by josdejong** [reply]
Thanks, that helps. Yes I think this is an Heroku issue indeed, I think I understand what's going on. 

For reference: I found the following discussion about this:

https://stackoverflow.com/questions/12704777/how-should-i-interpret-heroku-h18-errors

> The workaround is to always ensure the request body is read on the backend server before closing the connection.

and

> A POST/PUT request for files upload for example it must wait to read the whole body like this answer explains, So if you wish to disconnect a client sending a (very large) request body for exceeding file size limit - disconnecting the client will cause 503 error of heroku with HTML response of "Application Error". This is annoying if you build an API where client expect a JSON response, and for the example of file upload you can't even send 413 status code ONLY after processing the whole body (wasting time&bandwidth). The router does not respect the App server action of killing a connection.

I'll contact Heroku to ask about this.

**Reply by josdejong** [reply]
I've opened a ticket at Heroku asking about this issue. ~They do not see this as wrong behavior~ and keep repeating that ["entire body of the incoming request must be read"](https://devcenter.heroku.com/articles/error-codes#h18-server-request-interrupted) before sending a response. I still do not know why Heroku has this behavior.

I did some further research and according to the HTTP spec it is just fine to stop reading the request body in case of a 413 (see also [this stackoverflow discussion](https://stackoverflow.com/questions/18367824/how-to-cancel-http-upload-from-data-events/18370751#18370751)):

>    10.4.14 413 Request Entity Too Large
>
>    The server is refusing to process a request because the request
>    entity is larger than the server is willing or able to process. The
>    server MAY close the connection to prevent the client from continuing
>    the request.
>
>    https://www.rfc-editor.org/rfc/rfc2616#section-10.4.14

EDIT 2023-07-28: at first they kept repeating that the incoming request body must be read in full, but in the end they said they have a look at this issue. They could not give a timeline though.

**Reply by josdejong** [reply]
For now I implemented a Fastify plugin to flush the remaining data in case of a 413. 

@Uzlopak can you please have a short look on whether this is a solid solution? (This is my first Fastify project, I'm not very familiar with it).

```ts
// file: flushRemainingRequestData.ts

import http from 'http'
import fastifyPlugin from 'fastify-plugin'

/**
 * Make Heroku happy: flush all request data before responding with a 413.
 * Otherwise, Heroku will change the 413 in a 503 and log an H18 error.
 *
 * This plugin adds an onSend hook. When the response has a statusCode 413 and
 * the request is not complete, it will flush all leftover request data.
 *
 * Usage:
 *
 *     await myFastify.register(flushRemainingRequestData)
 *
 * See https://github.com/fastify/fastify/discussions/4919
 */
export const flushRemainingRequestData = fastifyPlugin(async (fastify) => {
  fastify.addHook('onSend', async (request, reply, payload) => {
    if (reply.statusCode === 413 && !request.raw.complete) {
      fastify.log.debug('About to send a 413 response. Flushing any left over request data...')

      const flushedBytes = await flushRequestData(request.raw)

      fastify.log.debug(`Flushed ${flushedBytes} bytes`)
    }

    return payload
  })
})

async function flushRequestData(rawRequest: http.IncomingMessage): Promise<number> {
  return new Promise((resolve, reject) => {
    let flushedBytes = 0

    rawRequest.on('data', onData)
    rawRequest.on('end', onEnd)
    rawRequest.on('error', onEnd)

    function onData(data: string) {
      flushedBytes += data.length
    }

    function onEnd(err: Error | undefined) {
      rawRequest.removeListener('data', onData)
      rawRequest.removeListener('end', onEnd)
      rawRequest.removeListener('error', onEnd)

      if (err) {
        reject(err)
      } else {
        resolve(flushedBytes)
      }
    }
  })
}
```

**Reply by metcoder95** [reply]
That should do the work, another thing you can do is iterate over the stream, pretty much like doing:

```js
const data = []
for await (const chunk of rawRequest) data.push(chunk)

console.log(Buffer.byteLength(Buffer.concat(data)))
```

The previous if you are interested in the size of the data.

If not really interested, another option is just allow the stream flow and wait for its end. e.g.

```js
const { once } = require('events');

// ...
rawRequest.resume();
await once(rawRequest, 'end');
```

**Reply by josdejong** [reply]
Using `resume()` is nice and compact! I indeed do not really need the byte count, that was just handy for debugging. Thanks Carlos.

Update from Heroku: at first they kept repeating that the incoming request body must be read in full, but in the end they said they have a look at this issue. They could not give a timeline though.

**Reply by Uzlopak** [reply]
Yeah. When i investigated this issue for you, i found an issue somewhere from 2017 or so. I would not expect any solution from heroku. Probably their infrastructure does not permit to reject requests early. And to fix this, they would need to invest much more resources than you just flushing the request into Nirvana.

**Reply by josdejong** [reply]
Yes that could very well be the case, we'll see. At least I have a functioning workaround now. Thanks for thinking along guys!

**Reply by Uzlopak** [reply]
What about creating a fastify plugin for heroku, where you always call resume() the raw request in the onSend hook. If the request is processed, than resume() should be not a big overhead. 
If this works without the statusCode check you could publish it on npm and we could list it as a community plugiin. So other devs just need to import it, and will not have the same ground work to do as you. Also if there are other quirks on heroku, they would report it to it.

Call it e.g. fastify-heroku

Or you connect with a dev from the heroku github org, after you implemented the plugin and propose to give them the project and npm package 

Then they could provide this as a solution, put it in their dev center etc..

**Reply by josdejong** [reply]
A plugin for Heroku could be useful indeed, there are some caveats getting that running.

The issues I came across so far getting a basic Fastify backend running on Heroku:

1. The application not working at all after being deployed on Heroku: you need to specify an IP `0.0.0.0`, see https://github.com/fastify/fastify/discussions/4104 
2. This very issue: Heroku turning a `413 Payload too large` into a `503 Service unavailable` and logging a H18 error
3. Not yet solved: when restarting via Heroku, the logs say "Process exited with status 143" which I don't understand: the server should gracefully exit with status 0.

I guess a plugin could help with that.

**Reply by DatoKhojava** [reply]
hey @josdejong  I use heroku free plain, if I post only required values from postman it works, but if I want to post whole document its gives me same error I even try to use host: 0.0.0.0 same error is there any idea ? 

[H18](https://github.com/hunterloftis/heroku-node-errcodes/tree/master/h18) here it says to use x-request-id, I get request_id from heroku logs but same its isn't working

**Reply by josdejong** [reply]
This 503 error is basically thrown when not all data sent by the client is read by the server, can that be the case?

**Reply by DatoKhojava** [reply]
in my case yes, coz server cant handle all data it can accept part of whole model. Can it be due to free plan? or maybe it needs to increase the size of received data??

**Reply by josdejong** [reply]
So did you implement a solution for this in your backend? Like I did by creating a fastify-plugin to flush all data: https://github.com/fastify/fastify/discussions/4919#discussioncomment-6553472

**Reply by DatoKhojava** [reply]
not yet ill try it but I don't have an positive expectations :DD but thank you anyway ill reply if its works for me
